### PR TITLE
:heart_eyes: show popup if a table does not exist

### PR DESCRIPTION
**Pull request by GabbasovDinar**
Step to reproduce: 

1. Open POS1 on device 1
2. Open POS2 on device 2
3. in POS1, use edit mode to create a new table, then save the table
4. POS2 has a Popup message

### CODE REVIEW

File: pos_multi_session/static/src/js/pos_multi_session.js [hunk 1]
```diff
@@ -335,28 +335,8 @@ odoo.define('pos_multi_session', function(require){
             var order = new models.Order({}, options);
             return order;
         },
-        ms_update_order: function(order, data){
+        ms_update_exist_order: function(order, data) {
```
Comment (yelizariev) [inline pos_multi_session/static/src/js/pos_multi_session.js]: existing

File: pos_multi_session_restaurant/static/src/js/pos_multi_session_restaurant.js [hunk 2]
```diff
@@ -95,13 +95,25 @@ odoo.define('pos_multi_session_restaurant', function(require){
...
+            }
+            if (table) {
+                var order = PosModelSuper.prototype.ms_create_order.apply(this, arguments);
                 order.table = self.tables_by_id[data.table_id];
                 order.customer_count = data.customer_count;
+                return order;
             }
-            return order;
+            this.gui.show_popup('confirm',{
+                'title': _t('Not found table'),
```
Comment (yelizariev) [inline pos_multi_session_restaurant/static/src/js/pos_multi_session_restaurant.js]: Got unknown table

**Review by yelizariev — APPROVED** [review]

**Comment by yelizariev** [comment]
My comments are fixed :+1:

**Review by Ramil-Mukhametzyanov — CHANGES_REQUESTED** [review]
(-) there is no toolbar to manage the configuration of the tables for demo user
(+) after the modification of the table configuration by the administrator and entering a table, a popup window appears for the demo user and the user can see the changes
(-) after the second modification, a pop-up window for admin appears
(+) but the latest changes for the administrator are still there
(-) demo user can operate on a deleted table
(-) the initialization of tables stops in the "Sync Order" process
The error in the browser console:
point_of_sale.assets.js:299 Uncaught TypeError: Cannot read property 'show' of undefined
	at Class.show_popup (point_of_sale.assets.js:299)
	at child.ms_create_order (point_of_sale.assets.js:661)
	at child.ms_update_order (point_of_sale.assets.js:603)
	at child.ms_update_order (point_of_sale.assets.js:664)
	at point_of_sale.assets.js:594
	at Function._.each._.forEach (web.assets_common.js:12)
	at child.updates_from_server (point_of_sale.assets.js:594)
	at child.updates_from_server (point_of_sale.assets.js:663)
	at Object.<anonymous> (point_of_sale.assets.js:650)
	at fire (web.assets_common.js:541)

Tested in pos-addons 757 runbot with updated and explicit modules

**Comment by GabbasovDinar** [comment]
@Ramil-Mukhametzyanov you can check the PR https://github.com/odoo/odoo/pull/27924 I think it is related

**Comment by Ramil-Mukhametzyanov** [comment]
> @Ramil-Mukhametzyanov you can check the PR [odoo/odoo#27924](https://github.com/odoo/odoo/pull/27924) I think it is related

The Odoo PR [odoo/odoo#27924] is closed, but nothing has changed.

**Review by Ramil-Mukhametzyanov — CHANGES_REQUESTED** [review]
The last changes didn't pass a test with two users admin and demo:
1. If demo user deletes a table and admin clicks on this table, pop-up "unknown table" appears for demo user, not for admin. ( it's the same crosswise)
2. The 'unknown' table displayed on the floor plan. (if we can detect it as unknown we can hide it after the click)
3. Deleting the first table causes continuous loop of errors. (orders are transfered on the first table, but not on the first free table.)

A table must be protected from deleting if it is only one table on the floor.

**Comment by GabbasovDinar** [comment]
> A table must be protected from deleting if it is only one table on the floor.

This question does not apply to this PR. In addition, you can always turn off the table edit setting for the user.

**Review by Ramil-Mukhametzyanov — CHANGES_REQUESTED** [review]
Almost perfect, but one thing should be fixed.

Steps:
1. Delete all the tables by an user
2. Click on a table by another user
Result:
The error 'transfer_table is undefined' in the first user's POS session.

It is necessary to protect the last table from deleting.

**Review by Ramil-Mukhametzyanov — APPROVED** [review]
:+1:

**Review by yelizariev — CHANGES_REQUESTED** [review]
No changelog available

File: pos_multi_session_restaurant/doc/changelog.rst [hunk 3]
```diff
@@ -1,3 +1,7 @@
+`3.2.5`
+-------
+- **IMP:** Synchronize the process of deleting or adding tables
```
Comment (yelizariev) [inline pos_multi_session_restaurant/doc/changelog.rst]: Use new tags for changelog

File: pos_multi_session/doc/changelog.rst [hunk 4]
```diff
@@ -1,3 +1,7 @@
+`4.2.3`
+-------
+
+- **IMP:** Code refactoring to support the "Sync restaurant orders" module
```
Comment (yelizariev) [inline pos_multi_session/doc/changelog.rst]: Use new tags

**Review by yelizariev — CHANGES_REQUESTED** [review]
Make just 3 commits for each update: 
*  pos_debt_notebook/doc/changelog.rst (lint)
* pos_multi_session
* pos_multi_session_restaurant

File: pos_multi_session/static/src/js/pos_multi_session.js [hunk 5]
```diff
@@ -334,37 +334,16 @@ odoo.define('pos_multi_session', function(require){
...
             }
         },
-        ms_create_order: function(options){
+        ms_create_order: function(options) {
+            var d = $.Deferred();
             options = _.extend({pos: this}, options || {});
             var order = new models.Order({}, options);
-            return order;
+            d.resolve(order);
+            return d;
```
Comment (yelizariev) [inline pos_multi_session/static/src/js/pos_multi_session.js]: Doubtful solution

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_multi_session/doc/changelog.rst [hunk 6]
```diff
@@ -1,130 +1,134 @@
+`4.2.3`
+-------
+
+**Improvement:** Code refactoring to support the "Sync restaurant orders" module
```
Comment (yelizariev) [inline pos_multi_session/doc/changelog.rst]: I don't see why this refactoring is needed

Comment (GabbasovDinar) [reply pos_multi_session/doc/changelog.rst]: @yelizariev Code refactoring to support **removing/adding tables** for the "Sync restaurant orders" module. Is it correctly?

**Pull request merged**